File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/base/NRange.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.base

import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asBoolean
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap

internal data class NRange<T : Number>(
    val min: T?,
    val max: T?,
    val includeMin: Boolean,
    val includeMax: Boolean,
) {
    fun isInRange(value: Number): Boolean {
        val minCheck = when {
            min == null -> true
            includeMin -> value.toDouble() >= min.toDouble()
            else -> value.toDouble() > min.toDouble()
        }

        val maxCheck = when {
            max == null -> true
            includeMax -> value.toDouble() <= max.toDouble()
            else -> value.toDouble() < max.toDouble()
        }

        return minCheck && maxCheck
    }

    companion object {
        fun fromMessageable(args: Any): NRange<out Number> = args.asMap().let { map ->
            val rawMin = map["min"]
            val rawMax = map["max"]
            val inMin = map["inMin"]!!.asBoolean()
            val inMax = map["inMax"]!!.asBoolean()

            when {
                rawMin is Int || rawMax is Int -> return NRange(
                    min = rawMin as Int?,
                    max = rawMax as Int?,
                    includeMin = inMin,
                    includeMax = inMax,
                )

                rawMin is Double || rawMax is Double -> return NRange(
                    min = rawMin as Double?,
                    max = rawMax as Double?,
                    includeMin = inMin,
                    includeMax = inMax,
                )

                else -> throw IllegalArgumentException("Invalid NRange type")
            }
        }

        inline fun <reified T : Number> fromMessageableWithExactType(args: Any): NRange<T> =
            args.asMap().let { map ->
                val rawMin = map["min"]
                val rawMax = map["max"]
                require(rawMin is T? && rawMax is T?)
                return NRange(
                    min = rawMin,
                    max = rawMax,
                    includeMin = map["inMin"]!!.asBoolean(),
                    includeMax = map["inMax"]!!.asBoolean(),
                )
            }
    }
}

File: ios/Classes/model/flutter_default_custom/NRange.swift
internal struct NRange<T: Numeric & Comparable & Hashable> : Hashable {
    let min: T?
    let max: T?
    let includeMin: Bool
    let includeMax: Bool

    func isInRange(_ value: T) -> Bool {
        let minCheck: Bool
        switch min {
        case .none:
            minCheck = true
        case .some(let minVal):
            minCheck = includeMin ? value.toDouble() >= minVal.toDouble() : value.toDouble() > minVal.toDouble()
        }

        let maxCheck: Bool
        switch max {
        case .none:
            maxCheck = true
        case .some(let maxVal):
            maxCheck = includeMax ? value.toDouble() <= maxVal.toDouble() : value.toDouble() < maxVal.toDouble()
        }

        return minCheck && maxCheck
    }
    
    static func fromMessageableWithExactType(args: Any) -> NRange<T> {
        guard let map = args as? [String: Any] else {
            fatalError("Invalid NRange type")
        }

        let rawMin = map["min"] as? T
        let rawMax = map["max"] as? T

        guard rawMin != nil && rawMax != nil else {
            fatalError("Invalid NRange type")
        }
        
        return NRange<T>(
            min: rawMin,
            max: rawMax,
            includeMin: map["inMin"] as! Bool,
            includeMax: map["inMax"] as! Bool
        )
    }
}

extension Numeric {
    func toDouble() -> Double {
        switch self {
        case let int as Int:
            return Double(int)
        case let double as Double:
            return double
        case let float as Float:
            return Double(float)
        default:
            fatalError("Unsupported Numeric type")
        }
    }
}
